### PR TITLE
New Meta data format for classification and object detction

### PR DESCRIPTION
**Pull request by swetha097**
Modify metadata structure for Classification and Object detection

### CODE REVIEW

File: rocAL/rocAL/include/meta_data/meta_data.h [hunk 1]
```diff
@@ -41,110 +41,144 @@ typedef struct BoundingBoxCord_
...
         _bb_cords_xcycwh =std::move(bb_cords_xcycwh);
-        _bb_label_ids = std::move(bb_label_ids);
-        _img_size = std::move(img_size);
+        _label_ids = std::move(bb_label_ids);
+        _info._img_size = std::move(img_size);
+        _info._img_id = std::move(img_id);
+    }
+    BoundingBoxCords& get_bb_cords() { return _bb_cords; }
+    BoundingBoxCords_xcycwh& get_bb_cords_xcycwh() { return _bb_cords_xcycwh; }
+    std::vector<size_t> get_bb_cords_dims()
```
Comment (fiona-gladwin) [inline rocAL/rocAL/include/meta_data/meta_data.h:137]: Check if there is a better way to get dims

File: rocAL/rocAL/include/meta_data/meta_data.h [hunk 2]
```diff
@@ -160,95 +194,105 @@ struct MetaDataBatch
         return this;
     }
```
Comment (fiona-gladwin) [inline rocAL/rocAL/include/meta_data/meta_data.h:195]: Need to check if the above concatenate and += operator is required, And remove it

File: rocAL/rocAL/include/meta_data/meta_data.h [hunk 3]
```diff
@@ -160,95 +194,105 @@ struct MetaDataBatch
...
-    std::vector<BoundingBoxCords_xcycwh>& get_bb_cords_batch_xcycxwh() { return _bb_cords_xcycwh; }
-    std::vector<BoundingBoxLabels>& get_bb_labels_batch() { return _bb_label_ids; }
-    ImgSizes & get_img_sizes_batch() { return _img_sizes; }
-    void reset_objects_count() { 
+    virtual std::vector<Labels>& get_label_batch() { };
+    virtual std::vector<BoundingBoxCords>& get_bb_cords_batch() { };
+    virtual std::vector<BoundingBoxCords_xcycwh>& get_bb_cords_batch_xcycxwh() { };
+    std::vector<uint>& get_image_id_batch() { return _info_batch._img_ids; }
+    std::vector<std::string>& get_image_names_batch() {return _info_batch._img_names; }
+    ImgSizes& get_img_sizes_batch() { return _info_batch._img_sizes; }
```
Comment (fiona-gladwin) [inline rocAL/rocAL/include/meta_data/meta_data.h:202]:
```suggestion
    std::vector<uint>& get_image_id_batch() { return _info_batch._img_ids; }
    std::vector<std::string>& get_image_names_batch() {return _info_batch._img_names; }
    ImgSizes& get_img_sizes_batch() { return _info_batch._img_sizes; }
```
The following 3 API can it be a part of the MetadataInfoBatch?

File: rocAL/rocAL/include/meta_data/meta_data.h [hunk 4]
```diff
@@ -160,95 +194,105 @@ struct MetaDataBatch
...
+    MetaDataDimensionsBatch& get_metadata_dimensions_batch() { return _metadata_dimensions; }
+    int get_batch_object_count() { return _total_objects_count; }
+    void reset_objects_count() {
         _total_objects_count = 0;
     }
     void increment_object_count(int count) { _total_objects_count += count; }
```
Comment (fiona-gladwin) [inline rocAL/rocAL/include/meta_data/meta_data.h:208]: Object count : Check if this is required, or if there is a better way to handle

File: rocAL/rocAL/include/meta_data/meta_data.h [hunk 5]
```diff
@@ -160,95 +194,105 @@ struct MetaDataBatch
...
+    MetaDataInfoBatch _info_batch;
+    int _total_objects_count = 0;
 };
 
 struct LabelBatch : public MetaDataBatch
 {
     void clear() override
     {
-        _label_id.clear();
+        for (int i = 0; i < _label_ids.size(); i++) {
```
Comment (fiona-gladwin) [inline rocAL/rocAL/include/meta_data/meta_data.h:219]: The can be changed to iterator?

File: rocAL/rocAL/include/meta_data/meta_data.h [hunk 6]
```diff
@@ -160,95 +194,105 @@ struct MetaDataBatch
...
+        _label_ids = std::move(labels);
     }
     LabelBatch() = default;
     void copy_data(std::vector<void*> buffer) override
     {
         if(buffer.size() < 1)
             THROW("The buffers are insufficient") // TODO -change
-        memcpy((int *)buffer[0], _label_id.data(), _label_id.size() * sizeof(int));
+        auto labels_buffer = (int *)buffer[0];
+        for (int i = 0; i < _label_ids.size(); i++) {
```
Comment (fiona-gladwin) [inline rocAL/rocAL/include/meta_data/meta_data.h:253]: This might delay the labels copy

File: rocAL/rocAL/include/meta_data/meta_data.h [hunk 7]
```diff
@@ -160,95 +194,105 @@ struct MetaDataBatch
...
+            memcpy(labels_buffer, _label_ids[i].data(), _label_ids[i].size() * sizeof(int));
+            labels_buffer += _label_ids[i].size();
+        }
     }
     std::vector<size_t>& get_buffer_size() override
```
Comment (fiona-gladwin) [inline rocAL/rocAL/include/meta_data/meta_data.h:258]: @fiona-gladwin Is this required?

File: rocAL/rocAL/include/meta_data/meta_data.h [hunk 8]
```diff
@@ -160,95 +194,105 @@ struct MetaDataBatch
...
         return *this;
     }
     void resize(int batch_size) override
     {
         _bb_cords.resize(batch_size);
-        _bb_label_ids.resize(batch_size);
-        _img_sizes.resize(batch_size);
+        _label_ids.resize(batch_size);
+        _info_batch._img_sizes.resize(batch_size);
+        _info_batch._img_ids.resize(batch_size);
```
Comment (fiona-gladwin) [inline rocAL/rocAL/include/meta_data/meta_data.h:295]: Should the resize for these happen here in this class?

File: rocAL/rocAL/include/meta_data/meta_data.h [hunk 9]
```diff
@@ -160,95 +194,105 @@ struct MetaDataBatch
...
+            memcpy(labels_buffer, _label_ids[i].data(), _label_ids[i].size() * sizeof(int));
+            labels_buffer += _label_ids[i].size();
+        }
     }
     std::vector<size_t>& get_buffer_size() override
     {
         _buffer_size.emplace_back(_total_objects_count * sizeof(int));
         return _buffer_size;
     }
+    std::vector<Labels>& get_label_batch() { return _label_ids; }
```
Comment (fiona-gladwin) [inline rocAL/rocAL/include/meta_data/meta_data.h:263]: get_labels_batch() would be better than get_label_batch()

File: rocAL/rocAL/source/meta_data/caffe2_meta_data_reader_detection.cpp [hunk 10]
```diff
@@ -74,25 +74,25 @@ void Caffe2MetaDataReaderDetection::lookup(const std::vector<std::string> &_imag
...
-    BoundingBoxLabels bb_labels;
+    Labels bb_labels;
 
     std::cerr << "\nMap contents: \n";
     for (auto &elem : _map_content)
     {
         std::cerr << "Name :\t " << elem.first;
         bb_coords = elem.second->get_bb_cords();
-        bb_labels = elem.second->get_bb_labels();
+        bb_labels = elem.second->get_label();
```
Comment (fiona-gladwin) [inline rocAL/rocAL/source/meta_data/caffe2_meta_data_reader_detection.cpp:95]: Change name of this API to get_labels()

File: rocAL/rocAL/include/meta_data/meta_data.h [hunk 11]
```diff
@@ -41,110 +41,144 @@ typedef struct BoundingBoxCord_
...
+    std::vector<size_t> _labels_dims = {};
+    int _object_count = 0;
 };
 
-struct BoundingBox : public MetaData
+struct BoundingBox : public Label
 {
     BoundingBox()= default;
-    BoundingBox(BoundingBoxCords bb_cords, BoundingBoxLabels bb_label_ids)
+    BoundingBox(BoundingBoxCords bb_cords, Labels bb_label_ids)
```
Comment (fiona-gladwin) [inline rocAL/rocAL/include/meta_data/meta_data.h:111]: Do we need multiple constructors?
Can't we have constructors with default values?

File: rocAL/rocAL/source/meta_data/caffe_meta_data_reader_detection.cpp [hunk 12]
```diff
@@ -178,7 +178,7 @@ void CaffeMetaDataReaderDetection::read_lmdb_record(std::string file_name, uint
 
                 bb_coords.push_back(box);
                 bb_labels.push_back(label);
-                add(file_name.c_str(), bb_coords, bb_labels, img_size);
+                add(file_name.c_str(), bb_coords, bb_labels, img_size, -1);
```
Comment (fiona-gladwin) [inline rocAL/rocAL/source/meta_data/caffe_meta_data_reader_detection.cpp:181]: Same comments as caffe2 detection, Do we need to pass image id explicitly as -1

File: rocAL/rocAL/source/meta_data/tf_meta_data_reader_detection.cpp [hunk 13]
```diff
@@ -187,14 +187,14 @@ void TFMetaDataReaderDetection::read_record(std::ifstream &file_contents, uint f
...
-      bbox_ymax = sf_ymax.float_list().value()[i]; 
+      bbox_ymax = sf_ymax.float_list().value()[i];
       box.l = bbox_xmin ;
       box.r = bbox_xmax ;
       box.t = bbox_ymin ;
       box.b = bbox_ymax ;
       bb_coords.push_back(box);
       bb_labels.push_back(label);
-      add(fname, bb_coords, bb_labels, img_size);
+      add(fname, bb_coords, bb_labels, img_size, -1);
```
Comment (fiona-gladwin) [inline rocAL/rocAL/source/meta_data/tf_meta_data_reader_detection.cpp:197]: Same comment applicable here

File: rocAL/rocAL/source/meta_data/coco_meta_data_reader.cpp [hunk 14]
```diff
@@ -127,11 +127,12 @@ void COCOMetaDataReader::read_all(const std::string &path)
     LookaheadParser parser(buff.get());
 
     BoundingBoxCords bb_coords;
-    BoundingBoxLabels bb_labels;
+    Labels bb_labels;
     ImgSizes img_sizes;
 
     BoundingBoxCord box;
     ImgSize img_size;
+    int original_id;
```
Comment (fiona-gladwin) [inline rocAL/rocAL/source/meta_data/coco_meta_data_reader.cpp:135]: Whys is this moved outside?

**Review by fiona-gladwin — COMMENTED** [review]
Change to be noted:
1. If image_id is -1 need not pass
2. Can multiple constructors be combined for Bbox
3. Rajy had asked to check if the Metadata API can return the references instead of values, need to check these

Some things to be checked:
1. Object count - is it required?
5. Better way to handle dims

File: rocAL/rocAL/include/meta_data/meta_data.h [hunk 15]
```diff
@@ -41,110 +41,144 @@ typedef struct BoundingBoxCord_
...
-    BoundingBoxCords _bb_cords = {}; // For bb use
-    BoundingBoxCords_xcycwh _bb_cords_xcycwh = {}; // For bb use
-    BoundingBoxLabels _bb_label_ids = {};// For bb use
-    ImgSize _img_size = {};
-    int _label_id = -1; // For label use only
-    std::vector<size_t> _bb_labels_dims = {};
-    std::vector<size_t> _bb_coords_dims = {};
-    int _object_count = 0;
+    virtual std::vector<int>& get_label() { };
+    virtual int get_object_count() { };
```
Comment (shobana-mcw) [inline rocAL/rocAL/include/meta_data/meta_data.h:58]: Is this wrt bbox? 
Then pls add bbox object count.

File: rocAL/rocAL/include/meta_data/meta_data.h [hunk 16]
```diff
@@ -41,110 +41,144 @@ typedef struct BoundingBoxCord_
...
+    virtual std::vector<size_t> get_label_dims() { };
```
Comment (shobana-mcw) [inline rocAL/rocAL/include/meta_data/meta_data.h:59]: What's the use case of get_label_dims?

File: rocAL/rocAL/include/meta_data/meta_data.h [hunk 17]
```diff
@@ -41,110 +41,144 @@ typedef struct BoundingBoxCord_
...
+    std::vector<int>& get_label() { return _label_ids; }
+    int get_object_count() { return _object_count; }
+    std::vector<size_t> get_label_dims()
+    {
+        _labels_dims = {_label_ids.size()};
+        return _labels_dims;
+    }
+    void set_labels(Labels label_ids)
+    {
+        _label_ids = std::move(label_ids);
```
Comment (shobana-mcw) [inline rocAL/rocAL/include/meta_data/meta_data.h:99]: Shouldn't we resize / reserve this variable?

File: rocAL/rocAL/include/meta_data/meta_data.h (hunk 3 above)
Comment (swetha097) [reply rocAL/rocAL/include/meta_data/meta_data.h:202]: Not needed for now

File: rocAL/rocAL/include/meta_data/meta_data.h [hunk 18]
```diff
@@ -41,110 +41,144 @@ typedef struct BoundingBoxCord_
...
 };
 
 struct Label : public MetaData
 {
     Label(int label)
     {
-        _label_id = label;
+        _label_ids.resize(1);
+        _label_ids[0] = label;
         _object_count = 1;
```
[the remark this answers is not in the document]
Comment (swetha097) [reply rocAL/rocAL/include/meta_data/meta_data.h:83]: Eliminated

File: rocAL/rocAL/include/meta_data/meta_data.h [hunk 19]
```diff
@@ -41,110 +41,144 @@ typedef struct BoundingBoxCord_
...
-struct BoundingBox : public MetaData
+struct BoundingBox : public Label
 {
     BoundingBox()= default;
-    BoundingBox(BoundingBoxCords bb_cords, BoundingBoxLabels bb_label_ids)
+    BoundingBox(BoundingBoxCords bb_cords, Labels bb_label_ids)
     {
         _bb_cords =std::move(bb_cords);
-        _bb_label_ids = std::move(bb_label_ids);
+        _label_ids = std::move(bb_label_ids);
```
[the remark this answers is not in the document]
Comment (swetha097) [reply rocAL/rocAL/include/meta_data/meta_data.h:114]: As per discussion, this is not required

File: rocAL/rocAL/include/meta_data/meta_data.h [hunk 20]
```diff
@@ -41,110 +41,144 @@ typedef struct BoundingBoxCord_
...
         _bb_cords =std::move(bb_cords);
-        _bb_label_ids = std::move(bb_label_ids);
-        _img_size = std::move(img_size);
+        _label_ids = std::move(bb_label_ids);
+        _info._img_size = std::move(img_size);
+        _info._img_id = std::move(img_id);
     }
-    void set_bb_cords(BoundingBoxCords bb_cords) { _bb_cords =std::move(bb_cords); }
-    BoundingBox(BoundingBoxCords_xcycwh bb_cords_xcycwh, BoundingBoxLabels bb_label_ids)
+    BoundingBox(BoundingBoxCords_xcycwh bb_cords_xcycwh, Labels bb_label_ids)
```
[the remark this answers is not in the document]
Comment (swetha097) [reply rocAL/rocAL/include/meta_data/meta_data.h:123]: @shobana-mcw Can you please explain the usage of this BoundingBoxCords_xcycwh and respective variable

File: rocAL/rocAL/include/meta_data/meta_data.h (hunk 7 above)
Comment (swetha097) [reply rocAL/rocAL/include/meta_data/meta_data.h:258]: Yes @fiona-gladwin. This was called by ring buffer before copy data called

File: rocAL/rocAL/include/meta_data/meta_data.h (hunk 15 above)
Comment (swetha097) [reply rocAL/rocAL/include/meta_data/meta_data.h]: get_object_count() is not required

File: rocAL/rocAL/include/meta_data/meta_data.h (hunk 16 above)
Comment (swetha097) [reply rocAL/rocAL/include/meta_data/meta_data.h]: It is removed

File: rocAL/rocAL/include/meta_data/meta_data.h (hunk 17 above)
Comment (swetha097) [reply rocAL/rocAL/include/meta_data/meta_data.h:89]: I think it not required @shobana-mcw

File: rocAL/rocAL/source/meta_data/coco_meta_data_reader.cpp (hunk 14 above)
Comment (swetha097) [reply rocAL/rocAL/source/meta_data/coco_meta_data_reader.cpp:133]: Moved inside

File: rocAL/rocAL/source/meta_data/coco_meta_data_reader.cpp [hunk 21]
```diff
@@ -50,48 +50,46 @@ void COCOMetaDataReader::lookup(const std::vector<std::string> &image_names)
...
-    _output->reset_objects_count();
     for (unsigned i = 0; i < image_names.size(); i++)
     {
         auto image_name = image_names[i];
         auto it = _map_content.find(image_name);
         if (_map_content.end() == it)
             THROW("ERROR: Given name not present in the map" + image_name)
         _output->get_bb_cords_batch()[i] = it->second->get_bb_cords();
-        _output->get_bb_labels_batch()[i] = it->second->get_bb_labels();
+        auto labels = it->second->get_labels();
```
Comment (fiona-gladwin) [inline rocAL/rocAL/source/meta_data/coco_meta_data_reader.cpp:60]: Is direct assignment not working here?

**Pull request closed without merging**